src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -52,67 +76,176 @@ public void AddDependencyLocation(string fullPath)
                 {
                     _knownAssemblyPathsBySimpleName[simpleName] = paths.Add(fullPath);
                 }
+
+                // Ensure that there is no cached Assembly information about this location. Long 
+                // lived processes like VS and VBCSCompiler will see the same location added as
+                // a dependency many times. Each time have to assume there is new content on disk
+                // that needs to be considered.
+                _analyzerAssemblyInfoMap[fullPath] = null;
```

Comment:
>_analyzerAssemblyInfoMap[fullPath] = null; [](http://example.com/codeflow?start=16&length=42)

Is the `AnalyzerAssemblyLoader` instance shared across multiple builds? If so, what is the impact of clearing `_analyzerAssemblyInfoMap[fullPath]` on other concurrent builds that may use the same dependency path?

Reply:
> Is the AnalyzerAssemblyLoader instance shared across multiple builds?

Yes.

> If so, what is the impact of clearing _analyzerAssemblyInfoMap[fullPath] on other concurrent builds that may use the same dependency path?

This is part of the second follow up item I have in #66532. The APIs are designed such that you can call `AddDependencyPath` followed by `LoadFromPath` in any order and combination. In truth though it only works 100% correct if you call them in a very specific pattern. Our uses of these APIs just so happen to use this pattern. 

The other invariant the loader has is that the on disk state cannot be modified in between calls to `AddDepedencyPath` and `LoadFromPath`. So even if two builds come along, call `AddDependencyPath` in parallel it's fine so long as this invariant is kept. One will clear the cached data, but the second will come along and put it back to the previous state. 

If you concurrent build, share analyzers, and change the on disk state of those locations it can result in quirky behavior. It has always been this way though. The previous code which attempted to cache assemblies was actually not doing what it thought it did. The underlying `AppDomain` and `AssemblyLoadContext` caches won out in all the load scenarios. 

I want to replace this with an APi which is *much* easier to follow. Compilations need to be registering their dependencies as a batch operation. That makes all of this clearer and afford us the ability to be more resilient in a few cases. But that's a follow up issue.